Mathlib/GroupTheory/Subsemigroup/Operations.lean

```diff
@@ -0,0 +1,899 @@
+theorem map_surjective_of_surjective : Function.Surjective (map f) :=
+  (giMapComap hf).l_surjective
+#align subsemigroup.map_surjective_of_surjective Subsemigroup.map_surjective_of_surjective
+
+@[to_additive]
+theorem comap_injective_of_surjective : Function.Injective (comap f) :=
+  (giMapComap hf).u_injective
+#align subsemigroup.comap_injective_of_surjective Subsemigroup.comap_injective_of_surjective
+
+@[to_additive]
+theorem map_inf_comap_of_surjective (S T : Subsemigroup N) :
+    (S.comap f ⊓ T.comap f).map f = S ⊓ T :=
+  (giMapComap hf).l_inf_u _ _
+#align subsemigroup.map_inf_comap_of_surjective Subsemigroup.map_inf_comap_of_surjective
+
+@[to_additive]
+theorem map_infᵢ_comap_of_surjective (S : ι → Subsemigroup N) :
+    (⨅ i, (S i).comap f).map f = infᵢ S :=
+  (giMapComap hf).l_infᵢ_u _
+#align subsemigroup.map_infi_comap_of_surjective Subsemigroup.map_infᵢ_comap_of_surjective
+
+@[to_additive]
+theorem map_sup_comap_of_surjective (S T : Subsemigroup N) :
+    (S.comap f ⊔ T.comap f).map f = S ⊔ T :=
+  (giMapComap hf).l_sup_u _ _
+#align subsemigroup.map_sup_comap_of_surjective Subsemigroup.map_sup_comap_of_surjective
+
+@[to_additive]
+theorem map_supᵢ_comap_of_surjective (S : ι → Subsemigroup N) :
+    (⨆ i, (S i).comap f).map f = supᵢ S :=
+  (giMapComap hf).l_supᵢ_u _
+#align subsemigroup.map_supr_comap_of_surjective Subsemigroup.map_supᵢ_comap_of_surjective
+
+@[to_additive]
+theorem comap_le_comap_iff_of_surjective {S T : Subsemigroup N} : S.comap f ≤ T.comap f ↔ S ≤ T :=
+  (giMapComap hf).u_le_u_iff
+#align subsemigroup.comap_le_comap_iff_of_surjective Subsemigroup.comap_le_comap_iff_of_surjective
+
+@[to_additive]
+theorem comap_strictMono_of_surjective : StrictMono (comap f) :=
+  (giMapComap hf).strictMono_u
+#align subsemigroup.comap_strict_mono_of_surjective Subsemigroup.comap_strictMono_of_surjective
+
+end GaloisInsertion
+
+end Subsemigroup
+
+namespace MulMemClass
+
+variable {A : Type _} [Mul M] [SetLike A M] [hA : MulMemClass A M] (S' : A)
+
+-- lower priority so other instances are found first
+/-- A submagma of a magma inherits a multiplication. -/
+@[to_additive "An additive submagma of an additive magma inherits an addition."]
+instance (priority := 900) hasMul : Mul S' :=
+  ⟨fun a b => ⟨a.1 * b.1, mul_mem a.2 b.2⟩⟩
+#align mul_mem_class.has_mul MulMemClass.hasMul
+
+-- lower priority so later simp lemmas are used first; to appease simp_nf
+@[simp, norm_cast, to_additive]
+theorem coe_mul (x y : S') : (↑(x * y) : M) = ↑x * ↑y :=
+  rfl
+#align mul_mem_class.coe_mul MulMemClass.coe_mul
+
+-- lower priority so later simp lemmas are used first; to appease simp_nf
+@[simp, to_additive]
+theorem mk_mul_mk (x y : M) (hx : x ∈ S') (hy : y ∈ S') :
+    (⟨x, hx⟩ : S') * ⟨y, hy⟩ = ⟨x * y, mul_mem hx hy⟩ :=
+  rfl
+#align mul_mem_class.mk_mul_mk MulMemClass.mk_mul_mk
+
+@[to_additive]
+theorem mul_def (x y : S') : x * y = ⟨x * y, mul_mem x.2 y.2⟩ :=
+  rfl
+#align mul_mem_class.mul_def MulMemClass.mul_def
```

Comment:
Should all the additive versions be `#align`ed, btw?

Reply:
This is all done